test_submodel.py

```diff
@@ -0,0 +1,85 @@
+# Example of how to construct a model as a series of submodels. Each
+# submodel is built, CoreNEURON data is written, and the submodel
+# is destroyed. This file can be run with
+#    python test_submodel.py -rparm
+# which writes data to the coredat folder. That folder can be compared to
+# the corenrn_data folder written by
+#    mpiexec -n 4 nrniv -python -mpi -rparm -coreneuron -filemode ringtest.py
+#    diff -r coredat corenrn_data
+# Note: -rparm is used for a more substantive test in that all the cells are
+# distinct with differt parameters.
+
+# The overall strategy for a ringtest model relies on the fact that it is
+# already does parallel setup so that there are pc.nhost() submodels,
+# one built on each rank, pc.id(). This setup did not use pc.nhost() or
+# pc.id() directly but stored those values in the global variables nhost and
+# rank respectively. So it is an easy matter to subvert that slightly and
+# run with a single process and iterate over range(nsubmodel) and for each
+# merely set rank and nhost to the proper values. The one exception to this
+# in the ringtest setup was the call to pc.set_gid2node(gid, rank) which was
+# changed to pc.set_gid2node(gid, pc.id()) since that function requires the
+# true rank of this process to function correctly. The other ringtest
+# transformation that was required was to factor out the ring build and
+# randomization into functions that are callable from here as well as in
+# the original ringtest.py .
+
+from neuron import h
+pc = h.ParallelContext()
+cvode = h.CVode()
+
+import ringtest
+
+def test_submodel(nsubmodel):
+  coredat = "./coredat"
+  cvode.cache_efficient(1)
+  gidgroups = [h.Vector() for _ in range(nsubmodel)] # used to write files.dat at end
+  for isubmodel in range(nsubmodel):
+    submodel = build_submodel(isubmodel, nsubmodel) # just like a single rank on an nhost cluster
+    pc.nrnbbcore_write("./coredat",  gidgroups[isubmodel])
+    teardown()
+    submodel = None
+
+    # verify no netcons or sections. Ready to go on to next submodel
+    assert (h.List("NetCon").count() == 0)
+    assert (len([s for s in h.allsec()]) == 0)
+
+  write_files_dat(coredat, gidgroups)
+
+def build_submodel(isubmodel, nsubmodel):
+  # fake nhost and rank
+  ringtest.settings.nhost = nsubmodel
+  ringtest.settings.rank = isubmodel
+
+  # broke into two parts to avoid timeit problems.
+  rings= ringtest.network()
+  ringtest.randomize(rings)
+
+  # same initialization as ringtest
+  pc.set_maxstep(10)
+  h.stdinit()
+
+  return rings
+
+def teardown():
+  pc.gid_clear()
+  # delete your NetCons list
+  # delete your Cells list
+  # unfortunately, cannot delete submodel here as there is reference to it
+  # in test_submodel(nsubmodel)
+
+# write out the files.dat file
+def write_files_dat(coredat, gidgroups):
+  f = open(coredat+"/files.dat", "w")
+  f.write("1.4\n") # CoreNEURON data version
```

Comment:
@nrnhines : can we get version as property or via some method? We have same problem of hard coding it in neurodamus and I wonder if it could be done in better way.

Reply:
I was thinking the same thing. The first thing that comes to mind is yet another ParallelContext method such as ```pc.nrncore_data_version()``` or perhaps a more generic thing such as ```pc.nrncore_property('property_name')```
since I would also like to know if coreneuron is available, the gpu is available, etc. Also the documentation for ```pc.nrnbbcore_write(...)``` is missing the first line data version requirement for files.dat.

Reply:
Or, as an end user, I don't want to manage combining these models like this. What I want to say is `take these subdir_1 subdir_2 subdir_3 and simulate it`. Internally we can do necessary symlinks and group files.dat.

Another possibility: coreneuron can be also updated to accept multiple data directories!  (I am now thinking **more** about this possibility!)

Reply:
We should discuss by zoom. I'm not seeing the value of multiple data directories for this problem since I can't envision the filenames not being unique regardless of the number of submodels.  You are right that files.dat can easily be handled by ```nrnbbcore_write``` but need to work out some api details. Maybe the user does not have to be aware of the gidgroups array of Vector either.

Reply:
I propose extending nrnbbcore_write with a new signature
```
pc.nrnbbcore_write([path, [i_submodel, n_submodel])
```
which indicates how many submodels are involved and at which point in the sequence this call is embedded. (when i_submodel == (nsubmodel - 1), the files.dat file is written at the end. Other possibilities are to count down from a beginning n_submodel to 0 and when 0 is reached, emit the files.dat file. In any event we would like for this to work on an mpi cluster and/or threads. It would also be nice to support the submodel strategy for the case of separate launch of NEURON for groups of submodels. That would require that the  accumulating groupgid information persist across launches which would mean writing some kind of intermediate file or else a files.dat to which nrnbbcore_write appends further gidgroup information as well as update the second line of the file (ngroup). If this latter is implemented then the signature could be further simplified to
```
pc.nrnbbcore_write([path], [bool append] )
```
which means that files.dat is to be updated at the end of this call and so is always valid when the call exits.
Lastly, I could imagine that the optional bool append could be eliminated and default is always append and it would be
up to the user to clear out the "outdat" folder or start a new one if wanting to start from the beginning.

Reply:
And is this a good time to allow the synonym, ```pc.nrncore_write([path], [bool append])```?
Note, in ```test_submodel.py``` this would change the statement to
```
pc.nrncore_write('./coredat', isubmodel != 0)
```
and would eliminate ```write_files_dat(...)```